src/storage/src/hummock/store/state_store.rs

```diff
@@ -137,6 +138,25 @@ impl HummockStorageCore {
             .iter(table_key_range, epoch, read_options, read_snapshot)
             .await
     }
+
+    pub async fn surely_not_have(
+        &self,
+        key_range: (Bound<Vec<u8>>, Bound<Vec<u8>>),
```

Comment:
Why we still need a `key_range` rather than a prefix.
I think this interface could only be used in some prefix value. We can not judge whether there exist value between some range.

Reply:
Seems that we are not only using the bloom filter to filter the key in this method. Before fetching the sst meta to get the bloom filter, we can filter by sst min-max with the key range.

Reply:
`state_store` could convert `prefix` to a `range` by itself. I think we shall support parameters as less as possible to avoid a wrong variable pass by upper layer.

Reply:
In our current code, all upper layer streaming executors are accessing storage state via `StateTable`, and it already provides a clean API, which only passes a pk_prefix, and it is already doing the work of converting the prefix to key range. 

On the other hand, no upper layer code of streaming executor is accessing `StateStore` directly, and we don't have to worry about misuse from the streaming executor side. Instead, I think we should provide better flexibility in the `StateStore` so that in the future when we want to support filtering other than the prefix filter in `StateTable`, we can easily reuse the current `StateStore` API without modifying it. 

Currently we are doing two kinds of filtering in our code, min-max and bloom filter, and they are independent to each other, and therefore it is more flexible to pass the two filtering condition (key range and bloom filter hint) in separate parameters. Passing a prefix solely will introduce the assumption that, all tables that call this method are using prefix as the bloom filter key. However, for some tables, they do not use the prefix as the bloom filter key.

Reply:
"However, for some tables, they do not use the prefix as the bloom filter key."
Can you give some specific examples?

Reply:
> Can you give some specific examples?

Did we previously have a plan to turn the `prefix_hint` into `bloom_filter_hint` in the `ReadOptions` so that we are not using a key prefix as the bloom filter key? IIRC, the demand for this feature comes from the earlier design to support watermark, which puts the timestamp column at the beginning of key encoding to support state clean by range delete, but build the bloom filter with the following distribution key.

Reply:
For watermark issue, we finally decide to decouple bloom filter key and distribution key, and now `bloom_filter_hint`/`bloom_filter_key` is always `pk_prefix`, and all streaming read will hit bloom filter.

Reply:
If we have decoupled bloom filter key and distribution key, I think we should pass the bloom filter key and the key range separately in the `StateStore` API 🤔. The `StateTable` will be responsible for calculating the bloom filter key according to the table config. Otherwise, they will be coupled again.

Reply:
> 

This is exactly the reason why I put `key_range` in this new interface.